File: LTSAgent.Server/LTSAgent.Frontend/UI/Input/ImagePicker.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using LTSAgent.Frontend.Infrastructure;

namespace LTSAgent.Frontend.UI.Input;

public partial class ImagePicker : JsComponentBase
{
    /// <summary> 이미지 변경 시 부모에 알리는 콜백 </summary>
    [Parameter] public EventCallback OnChanged { get; set; }

    /// <summary> 현재 첨부된 이미지의 MIME 타입 </summary>
    public string? ImageMediaType { get; private set; }

    /// <summary> 현재 첨부된 이미지의 Base64 데이터 </summary>
    public string? ImageBase64 { get; private set; }

    /// <summary> 파일 다이얼로그 참조 </summary>
    private ElementReference FileInputRef;

    /// <summary> .NET에서 JS가 호출할 수 있는 참조 </summary>
    private DotNetObjectReference<ImagePicker>? DotNetRef;

    protected override Task OnModuleLoaded()
    {
        DotNetRef = DotNetObjectReference.Create(this);
        return Task.CompletedTask;
    }

    /// <summary> 파일 다이얼로그를 오픈 </summary>
    private async Task Pick()
    {
        await Module.InvokeVoidAsync("pick", FileInputRef, DotNetRef);
    }

    /// <summary> JS에서 이미지 선택 완료 시 호출 </summary>
    [JSInvokable]
    public async Task OnImagePicked(string MediaType, string Base64)
    {
        ImageMediaType = MediaType;
        ImageBase64 = Base64;

        await OnChanged.InvokeAsync();
    }

    /// <summary> 첨부된 이미지를 제거 </summary>
    public async Task Clear()
    {
        ImageMediaType = null;
        ImageBase64 = null;
        
        await OnChanged.InvokeAsync();
    }
}

File: LTSAgent.Server/LTSAgent.Frontend/UI/Input/ImagePicker.razor.js
/**
 * 숨겨진 file input을 클릭하여 파일 다이얼로그를 엽니다.
 * 선택된 이미지를 Base64로 변환하여 C#에 전달합니다.
 */
export function pick(FileInputRef, dotNetRef)
{
    // 파일 선택시 호출되도록 함수 등록
    FileInputRef.onchange = function ()
    {
        var file = FileInputRef.files[0];
        if (!file)
            return;

        // 읽으면 호출되도록 함수 등록
        var reader = new FileReader();
        reader.onload = function ()
        {
            var base64 = reader.result.split(",")[1];
            var mimeType = file.type || "image/png";
            dotNetRef.invokeMethodAsync("OnImagePicked", mimeType, base64);
        };

        // 파일 읽기 시작
        reader.readAsDataURL(file);

        FileInputRef.value = "";
    };

    // 파일 다이얼로그 활성화
    FileInputRef.click();
}